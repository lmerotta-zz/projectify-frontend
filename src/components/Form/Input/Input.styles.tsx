import tw, { styled } from "twin.macro";
import { motion } from "framer-motion";

export const Wrapper = motion(
  tw.div`outline-none w-full flex flex-col items-stretch`
);

export const InputWrapper = motion(tw.div`w-full relative`);
export const Input = styled.input<{ invalid?: boolean }>`
  ${tw`w-full p-2 text-base border border-gray-400 bg-white appearance-none outline-none transition-colors focus:border-primary`}

  ${({ invalid }) => invalid && tw`border-danger!`}
`;
