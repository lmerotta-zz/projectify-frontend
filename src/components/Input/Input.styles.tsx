import tw, { styled } from "twin.macro";
import { motion } from "framer-motion";

export const Wrapper = motion(tw.div`relative outline-none`);

export const labelFilledStyles = tw`text-primary top-0 translate-y-0.5 text-xs`;

export const Label = styled.label`
  ${tw`absolute top-1/2 transform -translate-y-1/2 left-3 text-default cursor-text transition-all pointer-events-none`}
`;

export const InputWrapper = motion(tw.div`w-full relative`);
export const Input = styled.input<{ invalid?: boolean }>`
  ${tw`w-full pt-4 pb-2 pl-3 pr-3 border rounded border-gray-400 appearance-none outline-none transition-colors focus:border-primary`}
  &:focus {
    + .label {
      ${labelFilledStyles}
    }
  }

  :not(:placeholder-shown) {
    ${tw`border-primary`}
    + .label {
      ${labelFilledStyles}
    }
  }

  ${({ invalid }) => invalid && tw`border-red-600!`}
  + .label {
    ${({ invalid }) => invalid && tw`text-red-600!`}
  }
`;
