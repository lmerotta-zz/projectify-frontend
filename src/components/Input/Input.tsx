import { forwardRef, HTMLProps, ReactNode } from "react";
import tw, { styled } from "twin.macro";
import { AnimateSharedLayout, motion } from "framer-motion";
import ErrorMessage from "components/ErrorMessage/ErrorMessage";

const Wrapper = motion.custom(tw.div`relative outline-none`);

const labelFilledStyles = tw`text-primary top-0 translate-y-0.5 text-xs`;

const Label = styled.label`
  ${tw`absolute top-1/2 transform -translate-y-1/2 left-3 text-default cursor-text transition-all pointer-events-none`}
`;

const InputWrapper = motion.custom(tw.div`w-full relative`);
const StyledInput = styled.input`
  ${tw`w-full pt-4 pb-2 pl-3 pr-3 border rounded border-gray-400 appearance-none outline-none transition-colors focus:border-primary`}
  &:focus {
    + .label {
      ${labelFilledStyles}
    }
  }

  :not(:placeholder-shown) {
    ${tw`border-primary`}
    + .label {
      ${labelFilledStyles}
    }
  }

  &:invalid {
    ${tw`border-red-600`}
    + .label {
      ${tw`text-red-600`}
    }
  }
`;

type InputProps = Omit<HTMLProps<HTMLInputElement>, "as" | "name"> & {
  label: ReactNode;
  name: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ label, name, ...rest }, ref) => (
    <AnimateSharedLayout>
      <Wrapper layout>
        <InputWrapper layout>
          <StyledInput
            ref={ref}
            {...rest}
            name={name}
            placeholder="&nbsp;"
            data-testid={`input-${name}`}
          />
          {label && (
            <Label className="label" data-testid={`label-${name}`}>
              {label}
            </Label>
          )}
        </InputWrapper>
        <ErrorMessage name={name} />
      </Wrapper>
    </AnimateSharedLayout>
  )
);

export default Input;
