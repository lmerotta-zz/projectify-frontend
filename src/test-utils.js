import React from "react";
import { render } from "@testing-library/react";
import { CacheProvider } from "@emotion/react";
import createCache from "@emotion/cache";

const cache = createCache({ key: "iodajdoiwanoawdni" });
cache.compat = true;

const AllTheProviders = ({ children }) => {
  return <CacheProvider value={cache}>{children}</CacheProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
