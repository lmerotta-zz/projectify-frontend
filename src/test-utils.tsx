import { MockedProvider, MockedProviderProps } from "@apollo/client/testing";
import {
  CircularProgress,
  StyledEngineProvider,
  ThemeProvider,
} from "@mui/material";
import { render as realRender, RenderOptions } from "@testing-library/react";
import { ReactElement, ReactNode, Suspense } from "react";
import { MemoryRouter, MemoryRouterProps } from "react-router";
import { Route, Routes } from "react-router-dom";
import { ToastContainer } from "react-toastify";
import theme from "theme";

function renderWithProviders(
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries"> & {
    routerProps?: MemoryRouterProps;
    graphqlProps?: MockedProviderProps;
  }
) {
  const Wrapper = ({ children }: { children?: ReactNode }) => {
    return (
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <ToastContainer
            theme="colored"
            autoClose={3000}
            containerId="react-toastify"
          />
          <MockedProvider {...options?.graphqlProps} addTypename={false}>
            <MemoryRouter {...options?.routerProps}>
              <Suspense fallback={<CircularProgress />}>
                <Routes>
                  <Route
                    path="/*"
                    element={
                      <>
                        <h1>TESTSASATATAST</h1>
                        {children}
                      </>
                    }
                  />
                </Routes>
              </Suspense>
            </MemoryRouter>
          </MockedProvider>
        </ThemeProvider>
      </StyledEngineProvider>
    );
  };

  return realRender(ui, { wrapper: Wrapper, ...options });
}

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { renderWithProviders };
