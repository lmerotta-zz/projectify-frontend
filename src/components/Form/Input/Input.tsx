import { forwardRef, HTMLProps, ReactNode } from "react";
import * as Styles from "./Input.styles";
import ErrorMessage from "../ErrorMessage/ErrorMessage";
import { useFormContext } from "react-hook-form";

type InputProps = Omit<HTMLProps<HTMLInputElement>, "as" | "name"> & {
  label: ReactNode;
  name: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ label, name, ...rest }, ref) => {
    const { errors } = useFormContext();
    return (
      <Styles.Wrapper>
        <Styles.InputWrapper layout>
          <Styles.Input
            ref={ref}
            {...rest}
            name={name}
            placeholder="&nbsp;"
            data-testid={`input-${name}`}
            invalid={errors[name]?.length > 0}
          />
          {label && (
            <Styles.Label className="label" data-testid={`label-${name}`}>
              {label}
            </Styles.Label>
          )}
        </Styles.InputWrapper>
        <ErrorMessage name={name} />
      </Styles.Wrapper>
    );
  }
);

export default Input;
