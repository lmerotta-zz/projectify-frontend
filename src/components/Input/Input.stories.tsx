// Button.stories.tsx
import { ComponentProps, useEffect, useRef } from "react";
import Input from "./Input";
import { Story } from "@storybook/react/types-6-0";
import { FormProvider, useForm } from "react-hook-form";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Input",
  component: Input,
  argTypes: {
    label: { type: "string", defaultValue: "Label" },
    name: { defaultValue: "input-test" },
    hasError: { type: "boolean" },
  },
};

const Template: Story<ComponentProps<typeof Input> & { hasError?: string }> = ({
  hasError,
  name,
  ...args
}) => {
  const form = useForm();
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (hasError && !form.errors[name]) {
      form.setError(name, { message: "error message" });
      inputRef.current?.setCustomValidity("blep");
    }
  }, [hasError, form, name]);

  return (
    <FormProvider {...form}>
      <Input {...args} name={name} ref={inputRef} />
    </FormProvider>
  );
};

export const Default = Template.bind({});
export const WithError = Template.bind({});
WithError.args = {
  // @ts-ignore
  hasError: true,
};
