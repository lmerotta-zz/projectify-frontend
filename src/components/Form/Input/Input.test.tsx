import { render } from "test-utils";
import { useEffect } from "react";
import { FormProvider, useForm } from "react-hook-form";
import Input from "./Input";

describe("Input unit tests", () => {
  it("Applies focusing styles when input is focused", async () => {
    const Component = () => {
      return (
        <FormProvider {...useForm()}>
          <Input name="my-input" label="My input" />
        </FormProvider>
      );
    };
    const { asFragment } = render(<Component />);

    const firstRender = asFragment();
    expect(firstRender).toMatchSnapshot();
  });

  it("Applies the error styles", () => {
    const Component = () => {
      const form = useForm();

      useEffect(() => {
        form.setError("my-input", {
          type: "server",
          message: "Test message",
        });
      }, []);
      return (
        <FormProvider {...form}>
          <Input name="my-input" label="My input" />
        </FormProvider>
      );
    };

    const { getByTestId } = render(<Component />);

    expect(getByTestId("error-message-my-input")).toHaveTextContent(
      "Test message"
    );
  });
});
