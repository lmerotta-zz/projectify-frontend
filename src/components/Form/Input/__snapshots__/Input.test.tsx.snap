// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input unit tests Applies focusing styles when input is focused 1`] = `
<DocumentFragment>
  .emotion-0 {
  outline: 2px solid transparent;
  outline-offset: 2px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  width: 100%;
  position: relative;
}

.emotion-2 {
  width: 100%;
  padding: 0.5rem;
  font-size: 1rem;
  line-height: 1.5rem;
  border-width: 1px;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: 2px solid transparent;
  outline-offset: 2px;
  transition-property: background-color,border-color,color,fill,stroke;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
}

.emotion-2:focus {
  --tw-border-opacity: 1;
  border-color: rgba(147, 51, 234, var(--tw-border-opacity));
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <input
        class="emotion-2"
        data-testid="input-my-input"
        name="my-input"
        placeholder="Â "
      />
    </div>
  </div>
</DocumentFragment>
`;
